synapse/rest/client/room.py

```diff
@@ -1138,11 +1138,15 @@ async def on_POST(
 
 
 class RoomHierarchyRestServlet(RestServlet):
-    PATTERNS = (
+    PATTERNS = [
         re.compile(
             "^/_matrix/client/unstable/org.matrix.msc2946"
             "/rooms/(?P<room_id>[^/]*)/hierarchy$"
         ),
+    ] + list(
+        client_patterns(
+            "/rooms/(?P<room_id>[^/]*)/hierarchy$", releases=("v1",), unstable=False
+        )
     )
```

Comment:
ugh, I hate client_patterns. this is shorter and clearer to write out longhand.

Reply:
I'd prefer to tackle that separately if we wish to change it instead of tacking it onto an unrelated PR.

Reply:
it's not an unrelated PR - you're adding a new use of `client_patterns` when it would be shorter and clearer just to write it out.

But I'll not block this PR on it if you disagree.

Reply:
Do we have it documented that we should not be using `client_patterns` for new code? This is the first I've heard of this.

Reply:
> Do we have it documented that we should not be using client_patterns for new code?

That's not *quite* what I said - there may be times it's useful. What I *am* saying is that this:

```python
    PATTERNS = (
        re.compile(
            "^/_matrix/client/unstable/org.matrix.msc2946"
            "/rooms/(?P<room_id>[^/]*)/hierarchy$"
        ),
        re.compile(
            "^/_matrix/client/v1/rooms/(?P<room_id>[^/]*)/hierarchy$"
        ),
    )
```

is clearer and shorter than:

```python
    PATTERNS = [
        re.compile(
            "^/_matrix/client/unstable/org.matrix.msc2946"
            "/rooms/(?P<room_id>[^/]*)/hierarchy$"
        ),
    ] + list(
        client_patterns(
            "/rooms/(?P<room_id>[^/]*)/hierarchy$", releases=("v1",), unstable=False
        )
    )
```

... ie, I don't really feel like `client_patterns` is giving any value here. Just because it exists doesn't mean we should shoehorn it in everywhere we can.

Still, I'll not object if someone wants to go and rid the codebase of it :)

Reply:
> you can easily use client_patterns, which is the standard way we describe these URLs.

well, sometimes.

>  I would find it confusing to switch to client_patterns when removing the unstable API.

I would certainly not recommend doing that.